node/network/pov-distribution/src/lib.rs

```diff
@@ -561,45 +652,59 @@ impl PoVDistribution {
 	async fn run(
 		self,
 		mut ctx: impl SubsystemContext<Message = PoVDistributionMessage>,
-	) -> SubsystemResult<()> {
-		let mut state = State {
-			relay_parent_state: HashMap::new(),
-			peer_state: HashMap::new(),
-			our_view: View(Vec::new()),
-			metrics: self.metrics,
-		};
+	) -> error::Result<()> {
+		let mut state = State::default();
+		state.metrics = self.metrics;
 
 		loop {
-			match ctx.recv().await? {
-				FromOverseer::Signal(signal) => if handle_signal(&mut state, &mut ctx, signal).await? {
-					return Ok(());
-				},
-				FromOverseer::Communication { msg } => match msg {
-					PoVDistributionMessage::FetchPoV(relay_parent, descriptor, response_sender) =>
-						handle_fetch(
-							&mut state,
-							&mut ctx,
-							relay_parent,
-							descriptor,
-							response_sender,
-						).await?,
-					PoVDistributionMessage::DistributePoV(relay_parent, descriptor, pov) =>
-						handle_distribute(
-							&mut state,
-							&mut ctx,
-							relay_parent,
-							descriptor,
-							pov,
-						).await?,
-					PoVDistributionMessage::NetworkBridgeUpdateV1(event) =>
-						handle_network_update(
-							&mut state,
-							&mut ctx,
-							event,
-						).await?,
-				},
-			}
+			// `select_biased` is used since receiving connection notifications and
+			// peer view update messages may be racy and we want connection notifications
+			// first.
+			futures::select_biased! {
+				v = state.connection_requests.next().fuse() => {
+					match v {
+						Some((_relay_parent, _validator_id, peer_id)) => {
+							handle_validator_connected(&mut state, peer_id);
+						}
+						None => break,
+					}
+				}
+				v = ctx.recv().fuse() => {
```

Comment:
Here and above a call to `.fuse()` happens while `select_biased` is enclosed by a loop.

The docs for [`select_biased`](https://docs.rs/futures/0.3.8/futures/macro.select_biased.html) state the following:

> Futures and streams which are not already fused can be fused using the .fuse() method. Note, though, that fusing a future or stream directly in the call to select_biased! will not be enough to prevent it from being polled after completion if the select_biased! call is in a loop, so when select_biased!ing in a loop, users should take care to fuse() outside of the loop.

This seems to apply to this particular case, right?

Reply:
Long story short this happens. 

What `Fuse` is is this

```rust
pub struct Fuse<Fut>(#[pin] Option<Fut>);
```

(more or less the same for `Stream`). The idea is that we want a protection against polling a Future or a Stream that have finished.

The `connection_requests` field is a `Stream` so we would normally `.fuse()` the stream as the docs say before the loop and just be done with it. The problem with this is that `.fuse()` consumes `self` and we would have to move the `connection_requests` out of the `state` and `.fuse()` it. If that is done the code would no longer be able to update `connection_requests` as needed. So another approach is taken. Calling `.next()` on the `Stream` is essentially a `Future` that may be `await`ed or what not so we just want to create such a `Future` and `.fuse()` it here. It will be anyway thrown away after it becomes ready.

The same thing goes with the `recv()` which is also a `Future`. but this case is just a part of a previously described one. 

I will add a comment although I believe that this approach is not novel in the codebase: at least `collation-generation` and `subsystem-util` use the same one.

Reply:
You can just implement `FusedStream` for `ConnectionRequests` with `is_terminated` returning always `false`.

Reply:
Implementing `FusedStream` doesn't help.

As mentioned above, once you `.fuse()` a stream, you get back a stream-only interface. Whereas we want to call other methods on it, like `.put(..)` and `.remove(..)`.

Reply:
It does.